File: agentscope-service/frontend/src/api/orchestration.ts
/*
 * Copyright 2024-2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { api } from '@/lib/apiClient';

export type RunState = 'planned'|'running'|'waiting'|'paused'|'cancelling'|'cancelled'|'succeeded'|'partial_succeeded'|'failed';
export interface OrchestrationDefinition { id:string;tenant:string;namespace:string;name:string;description?:string;draftSpec:DefinitionSpec;draftVersion:number;version:number;createdAt:string;updatedAt:string;archivedAt?:string }
export interface DefinitionSpec { nodes:Array<Record<string,unknown>&{key:string;type:string}>;edges?:Array<{from:string;to:string;on?:string[];condition?:string;ordinal?:number}> }
export interface OrchestrationRevision { id:string;definitionId:string;revision:number;spec:DefinitionSpec;checksum:string;publishedAt:string }
export interface OrchestrationRun { id:string;tenant:string;namespace:string;rootIssueId:string;mode:string;definitionRevisionId?:string;parentRunId?:string;parentNodeId?:string;rerunOfRunId?:string;triggerType:string;input?:unknown;variables?:unknown;output?:unknown;policySnapshot?:unknown;usage?:unknown;state:RunState;waitReason?:string;failureCode?:string;failureMessage?:string;version:number;createdAt:string;startedAt?:string;completedAt?:string }
export interface RunNode { id:string;runId:string;nodeKey:string;definitionNodeKey?:string;type:string;role?:string;issueId?:string;state:string;config?:unknown;input?:unknown;output?:unknown;waitReason?:string;failureCode?:string;failureMessage?:string;version:number;startedAt?:string;completedAt?:string }
export interface RunEdge { id:string;runId:string;fromNodeId:string;toNodeId:string;onStates:string[];condition?:string;ordinal:number }
export interface ExecutionAttempt { id:string;agentTaskId:string;runId:string;nodeId:string;agentId?:string;bindingId?:string;attempt:number;dispatchGeneration:number;backendKind:string;state:string;runtimeBinding?:unknown;hostId?:string;agentInstanceId?:string;sessionId?:string;sessionRef?:string;turnId?:string;providerSessionId?:string;workspaceKey?:string;checkpoint?:unknown;result?:unknown;failureCode?:string;failureMessage?:string;usage?:unknown;createdAt:string;startedAt?:string;completedAt?:string }
export interface RunEvent { id:string;runId:string;sequence:number;nodeId?:string;agentTaskId?:string;attemptId?:string;type:string;actor:{type:string;ref?:string};payload?:unknown;occurredAt:string }
export interface RunGraph {childRuns?:OrchestrationRun[];definition?:OrchestrationDefinition;revision?:OrchestrationRevision;run:OrchestrationRun;nodes:RunNode[];edges:RunEdge[];tasks:Array<Record<string,unknown>&{id:string;agentId:string;status:string;runNodeId:string}>;attempts:ExecutionAttempt[]}

const q=(values:Record<string,string|undefined>)=>{const p=new URLSearchParams();Object.entries(values).forEach(([k,v])=>v&&p.set(k,v));const s=p.toString();return s?`?${s}`:''};
export const listDefinitions=(tenant:string,namespace:string)=>api.get<{definitions:OrchestrationDefinition[]}>(`/api/v1/orchestration-definitions${q({tenant,namespace})}`);
export const getDefinition=(id:string)=>api.get<{definition:OrchestrationDefinition}>(`/api/v1/orchestration-definitions/${id}`);
export const createDefinition=(body:unknown)=>api.post<{definition:OrchestrationDefinition}>('/api/v1/orchestration-definitions',body);
export const updateDefinition=(id:string,body:unknown)=>api.patch<{definition:OrchestrationDefinition}>(`/api/v1/orchestration-definitions/${id}`,body);
export const validateDefinition=(id:string,spec:DefinitionSpec)=>api.post<{valid:boolean;spec?:DefinitionSpec;error?:string}>(`/api/v1/orchestration-definitions/${id}/validate`,{spec});
export const publishDefinition=(id:string,expectedVersion?:number)=>api.post<{revision:OrchestrationRevision}>(`/api/v1/orchestration-definitions/${id}/publish`,{expectedVersion});
export const listRevisions=(id:string)=>api.get<{revisions:OrchestrationRevision[]}>(`/api/v1/orchestration-definitions/${id}/revisions`);
export const startRun=(id:string,body:unknown)=>api.post<{run:OrchestrationRun}>(`/api/v1/orchestration-definitions/${id}/runs`,body);
export const listRuns=(tenant:string,namespace:string,issueId?:string)=>api.get<{runs:OrchestrationRun[]}>(`/api/v1/orchestration-runs${q({tenant,namespace,issueId})}`);
export function normalizeRunGraph(graph: RunGraph): RunGraph { return {...graph, nodes:graph.nodes || [], edges:graph.edges || [], tasks:graph.tasks || [], attempts:graph.attempts || [], childRuns:graph.childRuns || []}; }
export const getRunGraph=async(id:string)=>normalizeRunGraph(await api.get<RunGraph>(`/api/v1/orchestration-runs/${id}/graph`));
export const listRunEvents=(id:string)=>api.get<{events:RunEvent[]}>(`/api/v1/orchestration-runs/${id}/events?limit=500`);
export const mutateRun=(id:string,action:'pause'|'resume'|'cancel')=>api.post<{run:OrchestrationRun}>(`/api/v1/orchestration-runs/${id}/${action}`,{});
export const rerunRun=(id:string)=>api.post<{run:OrchestrationRun}>(`/api/v1/orchestration-runs/${id}/rerun`,{idempotencyKey:crypto.randomUUID()});
export const signalRun=(id:string,name:string,payload:unknown)=>api.post<void>(`/api/v1/orchestration-runs/${id}/signals/${encodeURIComponent(name)}`,{idempotencyKey:crypto.randomUUID(),payload});
export const listAttempts=(tenant:string,namespace:string,taskId?:string)=>api.get<{attempts:ExecutionAttempt[]}>(`/api/v1/execution-attempts${q({tenant,namespace,taskId})}`);
export const listWorkflowRuns=(tenant:string,namespace:string,definitionId:string,offset=0)=>api.get<{runs:OrchestrationRun[]}>(`/api/v1/orchestration-runs${q({tenant,namespace,definitionId,offset:String(offset),limit:'100'})}`);
export const listWorkflowPage=(tenant:string,namespace:string,offset=0,archived=false)=>api.get<{definitions:OrchestrationDefinition[]}>(`/api/v1/orchestration-definitions${q({tenant,namespace,offset:String(offset),limit:'21',archived:String(archived)})}`);

File: agentscope-service/aistio/internal/controlplane/model/orchestration.go
// Copyright 2024-2026 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright 2024-2026 the original author or authors.
// Licensed under the Apache License, Version 2.0.

package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type OrchestrationRunMode string

const (
	RunModeDirect   OrchestrationRunMode = "direct"
	RunModeAdaptive OrchestrationRunMode = "adaptive"
	RunModeDeclared OrchestrationRunMode = "declared"
	RunModeSubrun   OrchestrationRunMode = "subrun"
)

type OrchestrationRunState string

const (
	RunPlanned          OrchestrationRunState = "planned"
	RunRunning          OrchestrationRunState = "running"
	RunWaiting          OrchestrationRunState = "waiting"
	RunPaused           OrchestrationRunState = "paused"
	RunCancelling       OrchestrationRunState = "cancelling"
	RunSucceeded        OrchestrationRunState = "succeeded"
	RunPartialSucceeded OrchestrationRunState = "partial_succeeded"
	RunFailed           OrchestrationRunState = "failed"
	RunCancelled        OrchestrationRunState = "cancelled"
)

func IsOrchestrationRunTerminal(state OrchestrationRunState) bool {
	return state == RunSucceeded || state == RunPartialSucceeded || state == RunFailed || state == RunCancelled
}

func CanTransitionOrchestrationRun(from, to OrchestrationRunState) bool {
	if from == to {
		return true
	}
	switch from {
	case RunPlanned:
		return to == RunRunning || to == RunCancelling || to == RunCancelled || to == RunFailed
	case RunRunning:
		return to == RunWaiting || to == RunPaused || to == RunCancelling || to == RunSucceeded || to == RunPartialSucceeded || to == RunFailed
	case RunWaiting:
		return to == RunRunning || to == RunPaused || to == RunCancelling || to == RunSucceeded || to == RunPartialSucceeded || to == RunFailed
	case RunPaused:
		return to == RunRunning || to == RunWaiting || to == RunCancelling
	case RunCancelling:
		return to == RunCancelled
	default:
		return false
	}
}

type RunNodeType string

const (
	RunNodeAgent     RunNodeType = "agent"
	RunNodeTeam      RunNodeType = "team"
	RunNodeApproval  RunNodeType = "approval"
	RunNodeCondition RunNodeType = "condition"
	RunNodeJoin      RunNodeType = "join"
	RunNodeTimer     RunNodeType = "timer"
	RunNodeSignal    RunNodeType = "signal"
	RunNodeSubrun    RunNodeType = "subrun"
)

type RunNodeState string

const (
	RunNodePending   RunNodeState = "pending"
	RunNodeReady     RunNodeState = "ready"
	RunNodeRunning   RunNodeState = "running"
	RunNodeWaiting   RunNodeState = "waiting"
	RunNodeSucceeded RunNodeState = "succeeded"
	RunNodeSkipped   RunNodeState = "skipped"
	RunNodeFailed    RunNodeState = "failed"
	RunNodeCancelled RunNodeState = "cancelled"
)

func IsRunNodeTerminal(state RunNodeState) bool {
	return state == RunNodeSucceeded || state == RunNodeSkipped || state == RunNodeFailed || state == RunNodeCancelled
}

func CanTransitionRunNode(from, to RunNodeState) bool {
	if from == to {
		return true
	}
	switch from {
	case RunNodePending:
		return to == RunNodeReady || to == RunNodeSkipped || to == RunNodeCancelled
	case RunNodeReady:
		return to == RunNodeRunning || to == RunNodeWaiting || to == RunNodeSkipped || to == RunNodeFailed || to == RunNodeCancelled
	case RunNodeRunning:
		return to == RunNodeWaiting || to == RunNodeSucceeded || to == RunNodeFailed || to == RunNodeCancelled
	case RunNodeWaiting:
		return to == RunNodeReady || to == RunNodeRunning || to == RunNodeSucceeded || to == RunNodeFailed || to == RunNodeCancelled
	default:
		return false
	}
}

type OrchestrationDefinition struct {
	ID           uuid.UUID       `json:"id"`
	Tenant       string          `json:"tenant"`
	Namespace    string          `json:"namespace"`
	Name         string          `json:"name"`
	Description  string          `json:"description,omitempty"`
	DraftSpec    json.RawMessage `json:"draftSpec"`
	DraftVersion int64           `json:"draftVersion"`
	Version      int64           `json:"version"`
	CreatedBy    Actor           `json:"createdBy"`
	CreatedAt    time.Time       `json:"createdAt"`
	UpdatedAt    time.Time       `json:"updatedAt"`
	ArchivedAt   *time.Time      `json:"archivedAt,omitempty"`
}

type OrchestrationRevision struct {
	ExpectedDefinitionVersion int64           `json:"-"`
	ID                        uuid.UUID       `json:"id"`
	DefinitionID              uuid.UUID       `json:"definitionId"`
	Tenant                    string          `json:"tenant"`
	Namespace                 string          `json:"namespace"`
	Revision                  int64           `json:"revision"`
	Spec                      json.RawMessage `json:"spec"`
	Checksum                  string          `json:"checksum"`
	PublishedBy               Actor           `json:"publishedBy"`
	PublishedAt               time.Time       `json:"publishedAt"`
}

type OrchestrationRun struct {
	ID                   uuid.UUID             `json:"id"`
	Tenant               string                `json:"tenant"`
	Namespace            string                `json:"namespace"`
	RootIssueID          uuid.UUID             `json:"rootIssueId"`
	Mode                 OrchestrationRunMode  `json:"mode"`
	DefinitionRevisionID *uuid.UUID            `json:"definitionRevisionId,omitempty"`
	ParentRunID          *uuid.UUID            `json:"parentRunId,omitempty"`
	ParentNodeID         *uuid.UUID            `json:"parentNodeId,omitempty"`
	RerunOfRunID         *uuid.UUID            `json:"rerunOfRunId,omitempty"`
	TriggerType          string                `json:"triggerType"`
	TriggerRef           string                `json:"triggerRef,omitempty"`
	IdempotencyKey       string                `json:"idempotencyKey,omitempty"`
	Input                json.RawMessage       `json:"input,omitempty"`
	Variables            json.RawMessage       `json:"variables,omitempty"`
	Output               json.RawMessage       `json:"output,omitempty"`
	PolicySnapshot       json.RawMessage       `json:"policySnapshot,omitempty"`
	Usage                json.RawMessage       `json:"usage,omitempty"`
	State                OrchestrationRunState `json:"state"`
	WaitReason           string                `json:"waitReason,omitempty"`
	FailureCode          string                `json:"failureCode,omitempty"`
	FailureMessage       string                `json:"failureMessage,omitempty"`
	Version              int64                 `json:"version"`
	CreatedBy            Actor                 `json:"createdBy"`
	CreatedAt            time.Time             `json:"createdAt"`
	UpdatedAt            time.Time             `json:"updatedAt"`
	StartedAt            *time.Time            `json:"startedAt,omitempty"`
	CompletedAt          *time.Time            `json:"completedAt,omitempty"`
}

type RunNode struct {
	ID                uuid.UUID       `json:"id"`
	RunID             uuid.UUID       `json:"runId"`
	Tenant            string          `json:"tenant"`
	Namespace         string          `json:"namespace"`
	NodeKey           string          `json:"nodeKey"`
	DefinitionNodeKey string          `json:"definitionNodeKey,omitempty"`
	Type              RunNodeType     `json:"type"`
	Role              string          `json:"role,omitempty"`
	IssueID           *uuid.UUID      `json:"issueId,omitempty"`
	State             RunNodeState    `json:"state"`
	Config            json.RawMessage `json:"config,omitempty"`
	Input             json.RawMessage `json:"input,omitempty"`
	Output            json.RawMessage `json:"output,omitempty"`
	Iteration         int32           `json:"iteration"`
	WaitReason        string          `json:"waitReason,omitempty"`
	FailureCode       string          `json:"failureCode,omitempty"`
	FailureMessage    string          `json:"failureMessage,omitempty"`
	Version           int64           `json:"version"`
	CreatedAt         time.Time       `json:"createdAt"`
	UpdatedAt         time.Time       `json:"updatedAt"`
	StartedAt         *time.Time      `json:"startedAt,omitempty"`
	CompletedAt       *time.Time      `json:"completedAt,omitempty"`
}

type RunEdge struct {
	ID         uuid.UUID       `json:"id"`
	RunID      uuid.UUID       `json:"runId"`
	Tenant     string          `json:"tenant"`
	Namespace  string          `json:"namespace"`
	FromNodeID uuid.UUID       `json:"fromNodeId"`
	ToNodeID   uuid.UUID       `json:"toNodeId"`
	OnStates   []RunNodeState  `json:"onStates"`
	Condition  string          `json:"condition,omitempty"`
	Ordinal    int32           `json:"ordinal"`
	Metadata   json.RawMessage `json:"metadata,omitempty"`
	CreatedAt  time.Time       `json:"createdAt"`
}

type RunTeamSnapshot struct {
	RunID     uuid.UUID       `json:"runId"`
	TeamID    uuid.UUID       `json:"teamId"`
	Tenant    string          `json:"tenant"`
	Namespace string          `json:"namespace"`
	Snapshot  json.RawMessage `json:"snapshot"`
	CreatedAt time.Time       `json:"createdAt"`
}

type RunEvent struct {
	ID             uuid.UUID       `json:"id"`
	RunID          uuid.UUID       `json:"runId"`
	Tenant         string          `json:"tenant"`
	Namespace      string          `json:"namespace"`
	Sequence       int64           `json:"sequence"`
	NodeID         *uuid.UUID      `json:"nodeId,omitempty"`
	AgentTaskID    *uuid.UUID      `json:"agentTaskId,omitempty"`
	AttemptID      *uuid.UUID      `json:"attemptId,omitempty"`
	Type           string          `json:"type"`
	Actor          Actor           `json:"actor"`
	Payload        json.RawMessage `json:"payload,omitempty"`
	CausationID    string          `json:"causationId,omitempty"`
	CorrelationID  string          `json:"correlationId,omitempty"`
	IdempotencyKey string          `json:"idempotencyKey,omitempty"`
	OccurredAt     time.Time       `json:"occurredAt"`
}

type RuntimeBindingCandidate struct {
	Binding              RuntimeBinding  `json:"binding"`
	RequiredCapabilities json.RawMessage `json:"requiredCapabilities,omitempty"`
	SecurityConstraints  json.RawMessage `json:"securityConstraints,omitempty"`
	// Dispatch metadata is assigned by the scheduler and frozen into the task.
	SelectionSource string `json:"selectionSource,omitempty"`
	CandidateIndex  int32  `json:"candidateIndex,omitempty"`
}

// RuntimeBindingPolicy is the reusable ordered candidate policy embedded in a
// Team member snapshot. AgentRuntimePolicy adds persistence identity,
// concurrency, and timeout controls around the same selection semantics.
type RuntimeBindingPolicy struct {
	Candidates    []RuntimeBindingCandidate `json:"candidates"`
	SelectionMode string                    `json:"selectionMode"`
	FallbackMode  string                    `json:"fallbackMode"`
	RetryPolicy   json.RawMessage           `json:"retryPolicy,omitempty"`
}

type AgentRuntimePolicy struct {
	ID                    uuid.UUID                 `json:"id"`
	Tenant                string                    `json:"tenant"`
	Namespace             string                    `json:"namespace"`
	AgentRef              string                    `json:"agentId"`
	Candidates            []RuntimeBindingCandidate `json:"candidates"`
	SelectionMode         string                    `json:"selectionMode"`
	FallbackMode          string                    `json:"fallbackMode"`
	MaxConcurrency        int32                     `json:"maxConcurrency,omitempty"`
	QueueTimeoutSeconds   int64                     `json:"queueTimeoutSeconds,omitempty"`
	AttemptTimeoutSeconds int64                     `json:"attemptTimeoutSeconds,omitempty"`
	RetryPolicy           json.RawMessage           `json:"retryPolicy,omitempty"`
	Version               int64                     `json:"version"`
	CreatedAt             time.Time                 `json:"createdAt"`
	UpdatedAt             time.Time                 `json:"updatedAt"`
	ArchivedAt            *time.Time                `json:"archivedAt,omitempty"`
}
